### ISSUE TITLE
echo_reply ignored when coming in on different interface

### ISSUE THREAD

**Issue by sakul95** [zmap/zmap#633]
**Describe the bug**
When starting a ZMap scan with icmp_echoscan over a certain interface with given gateway-mac and source-ip, it can happen, that the echo_reply comes in on an other interface than the echo_request was sent.
Then ZMap does not recognize this answer and it appears not to be reachable.
Trying to reach the same ip addresses with ping is possible, even if the echo_reply comes in over the "wrong" interface.

**CLI Arguments**
ZMap only displays "193.169.204.70" as reachable, "177.125.150.95" seems to be lost.
Ping can ping both.

```
sudo zmap -i tap5 -r 100 -M icmp_echoscan --source-ip=184.164.235.1 -I /tmp/hitlisttest_v4 --gateway-mac=c2:39:de:b9:c2:ec

ping4 -I 184.164.235.1 193.169.204.70
ping4 -I 184.164.235.1 177.125.150.95

cat /tmp/hitlisttest_v4
193.169.204.70
177.125.150.95

sudo tcpdump -e -i any host 177.125.150.95 or host 193.169.204.70
TCPDUMP while PING
11:41:34.966440 Out 5e:33:ae:36:5e:37 (oui Unknown) ethertype IPv4 (0x0800), length 100: 184.164.235.1 > 193.169.204.70: ICMP echo request, id 1251, seq 1, length 64
11:41:35.000138  In c2:39:de:b9:c2:ec (oui Unknown) ethertype IPv4 (0x0800), length 100: 193.169.204.70 > 184.164.235.1: ICMP echo reply, id 1251, seq 1, length 64

11:41:41.903974 Out b6:d0:72:29:c7:59 (oui Unknown) ethertype IPv4 (0x0800), length 100: 184.164.235.1 > 177.125.150.95: ICMP echo request, id 1269, seq 1, length 64
11:41:42.216155  In f6:ad:dc:cd:60:7f (oui Unknown) ethertype IPv4 (0x0800), length 100: 177.125.150.95 > 184.164.235.1: ICMP echo reply, id 1269, seq 1, length 64

TCPDUMP while ZMap
11:16:04.721532  In c2:39:de:b9:c2:ec (oui Unknown) ethertype IPv4 (0x0800), length 56: 193.169.204.70 > 184.164.235.1: ICMP echo reply, id 35397, seq 40100, length 20
11:17:03.358541  In f6:ad:dc:cd:60:7f (oui Unknown) ethertype IPv4 (0x0800), length 56: 177.125.150.95 > 184.164.235.1: ICMP echo reply, id 21148, seq 27869, length 20
```

**Example Target IP**

193.169.204.70
177.125.150.95

**Expected behavior**
No hit in ZMap for an IP address, if the reply comes in over another interface than the request was sent from.

**Environment:**
 - OS: Debian GNU/Linux 10 (buster)
 - Version: zmap Development Build. Commit dd31295 - Sat Aug 24 20:58:25 2019 +0200

**Additional context**
Is it possible to add an parameter that tells ZMap to listen to all interfaces, if you consider this as a feature, not a bug?

**Comment by zakird** [comment]
Can you see if https://github.com/zmap/zmap/pull/635 fixes the problem?

**Comment by sakul95** [comment]
First of all, thanks for your help! 
I could not verify it because of an error in the installation process:
```
[ 64%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_icmp_echo.c.o
/opt/zmap/zmap/src/zmap.c: In function ‘main’:
/opt/zmap/zmap/src/zmap.c:464:10: error: ‘struct gengetopt_args_info’ has no member named ‘listen_all_interfaces’
  if (args.listen_all_interfaces) {
          ^
make[2]: *** [src/CMakeFiles/zmap.dir/build.make:283: src/CMakeFiles/zmap.dir/zmap.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:226: src/CMakeFiles/zmap.dir/all] Error 2
```

<details>
  <summary>Complete Error:</summary>
  
  ```
lukas@machine:/opt/zmap$ git clone https://github.com/zmap/zmap.git
Cloning into 'zmap'...
remote: Enumerating objects: 34, done.
remote: Counting objects: 100% (34/34), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 6704 (delta 10), reused 11 (delta 4), pack-reused 6670
Receiving objects: 100% (6704/6704), 5.93 MiB | 1.01 MiB/s, done.
Resolving deltas: 100% (4737/4737), done.
lukas@machine:/opt/zmap$ ls
target_file  zmap
lukas@machine:/opt/zmap$ cd zmap/
lukas@machine:/opt/zmap/zmap$ git checkout 0d279da65e4f5f28c91ebf3cf151e2950d5640e5
Note: checking out '0d279da65e4f5f28c91ebf3cf151e2950d5640e5'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 0d279da allow listening on all interfaces
lukas@machine:/opt/zmap/zmap$ cmake .
-- The C compiler identification is GNU 8.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29") 
-- Checking for module 'json-c'
--   Found json-c, version 0.12.1
-- Default ZMap configuration file location is /etc/zmap
-- Checking if blocklist.conf exists there...
-- Checking if zmap.conf exists there...
CMake Warning at CMakeLists.txt:144 (message):
  Existing configuration file detected at /etc/zmap/zmap.conf, zmap.conf from
  sources will NOT be installed.  Please check and install manually!


-- Configuring done
-- Generating done
-- Build files have been written to: /opt/zmap/zmap
lukas@machine:/opt/zmap/zmap$ make -j4
Scanning dependencies of target git_versioning
Scanning dependencies of target zmaplib
[  1%] Building C object lib/CMakeFiles/zmaplib.dir/blocklist.c.o
[  2%] Building C object lib/CMakeFiles/zmaplib.dir/constraint.c.o
[  3%] Building C object lib/CMakeFiles/zmaplib.dir/logger.c.o
[  3%] Built target git_versioning
[  4%] Building C object lib/CMakeFiles/zmaplib.dir/pbm.c.o
[  5%] Building C object lib/CMakeFiles/zmaplib.dir/random.c.o
[  6%] Building C object lib/CMakeFiles/zmaplib.dir/rijndael-alg-fst.c.o
[  6%] Building C object lib/CMakeFiles/zmaplib.dir/xalloc.c.o
[  7%] Building C object lib/CMakeFiles/zmaplib.dir/lockfd.c.o
[  8%] Building C object lib/CMakeFiles/zmaplib.dir/util.c.o
[  9%] Building C object lib/CMakeFiles/zmaplib.dir/queue.c.o
[ 10%] Building C object lib/CMakeFiles/zmaplib.dir/csv.c.o
[ 11%] Linking C static library libzmaplib.a
[ 11%] Built target zmaplib
[ 12%] Generating zitopt.h
[ 12%] Generating zbopt.h
[ 12%] Generating parser.c
[ 13%] Generating parser.c
[ 15%] Generating ztopt.h
[ 15%] Generating zopt.h
Scanning dependencies of target ziterate
Scanning dependencies of target zblocklist
[ 16%] Generating lexer.c
[ 17%] Generating lexer.c
[ 18%] Building C object src/CMakeFiles/zblocklist.dir/zblocklist.c.o
[ 19%] Building C object src/CMakeFiles/ziterate.dir/aesrand.c.o
[ 20%] Building C object src/CMakeFiles/ziterate.dir/cyclic.c.o
[ 21%] Building C object src/CMakeFiles/zblocklist.dir/zbopt_compat.c.o
In file included from /opt/zmap/zmap/src/zbopt_compat.c:17:
/opt/zmap/zmap/src/zbopt.c: In function ‘cmdline_parser_internal’:
/opt/zmap/zmap/src/zbopt.c:505:7: warning: variable ‘check_required’ set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 22%] Building C object src/CMakeFiles/ziterate.dir/iterator.c.o
Scanning dependencies of target zmap
Scanning dependencies of target ztests
[ 23%] Building C object src/CMakeFiles/zmap.dir/aesrand.c.o
[ 24%] Building C object src/CMakeFiles/ztests.dir/aesrand.c.o
[ 25%] Building C object src/CMakeFiles/ziterate.dir/shard.c.o
[ 27%] Building C object src/CMakeFiles/zmap.dir/cyclic.c.o
[ 27%] Linking C executable zblocklist
[ 28%] Building C object src/CMakeFiles/ztests.dir/cyclic.c.o
[ 29%] Building C object src/CMakeFiles/ziterate.dir/state.c.o
[ 29%] Built target zblocklist
/opt/zmap/zmap/src/state.c:45:30: warning: initialization of ‘unsigned int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
       .source_ip_addresses = NULL,
                              ^~~~
/opt/zmap/zmap/src/state.c:45:30: note: (near initialization for ‘zconf.source_ip_addresses[0]’)
/opt/zmap/zmap/src/state.c:13:27: warning: missing braces around initializer [-Wmissing-braces]
 struct state_conf zconf = {.log_level = LOG_INFO,
                           ^
[ 30%] Generating topt.h
[ 31%] Building C object src/CMakeFiles/zmap.dir/expression.c.o
Scanning dependencies of target ztee
[ 32%] Building C object src/CMakeFiles/ziterate.dir/validate.c.o
[ 33%] Building C object src/CMakeFiles/ztests.dir/expression.c.o
[ 34%] Building C object src/CMakeFiles/ztee.dir/ztee.c.o
[ 35%] Building C object src/CMakeFiles/ziterate.dir/zitopt_compat.c.o
[ 36%] Building C object src/CMakeFiles/zmap.dir/fieldset.c.o
[ 37%] Building C object src/CMakeFiles/ztests.dir/fieldset.c.o
In file included from /opt/zmap/zmap/src/zitopt_compat.c:17:
/opt/zmap/zmap/src/zitopt.c: In function ‘cmdline_parser_internal’:
/opt/zmap/zmap/src/zitopt.c:545:7: warning: variable ‘check_required’ set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 38%] Building C object src/CMakeFiles/ztee.dir/topt_compat.c.o
In file included from /opt/zmap/zmap/src/topt_compat.c:9:
/opt/zmap/zmap/src/topt.c: In function ‘update_arg’:
/opt/zmap/zmap/src/topt.c:397:9: warning: variable ‘stop_char’ set but not used [-Wunused-but-set-variable]
   char *stop_char = 0;
         ^~~~~~~~~
/opt/zmap/zmap/src/topt.c: In function ‘cmdline_parser_internal’:
/opt/zmap/zmap/src/topt.c:476:7: warning: variable ‘check_required’ set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 39%] Building C object src/CMakeFiles/zmap.dir/filter.c.o
[ 40%] Building C object src/CMakeFiles/ztests.dir/filter.c.o
[ 40%] Building C object src/CMakeFiles/ziterate.dir/ziterate.c.o
[ 41%] Linking C executable ztee
[ 43%] Building C object src/CMakeFiles/zmap.dir/get_gateway.c.o
[ 43%] Building C object src/CMakeFiles/ztests.dir/get_gateway.c.o
[ 44%] Linking C executable ziterate
[ 44%] Built target ztee
[ 45%] Building C object src/CMakeFiles/ztests.dir/iterator.c.o
[ 45%] Built target ziterate
[ 46%] Building C object src/CMakeFiles/zmap.dir/iterator.c.o
[ 46%] Building C object src/CMakeFiles/zmap.dir/monitor.c.o
[ 47%] Building C object src/CMakeFiles/zmap.dir/recv.c.o
[ 48%] Building C object src/CMakeFiles/ztests.dir/monitor.c.o
/opt/zmap/zmap/src/monitor.c: In function ‘onscreen_appsuccess’:
/opt/zmap/zmap/src/monitor.c:312:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; sent: %llu %sp/s (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:319:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c: In function ‘onscreen_appsuccess’:
/opt/zmap/zmap/src/monitor.c:312:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; sent: %llu %sp/s (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:319:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:328:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; sent: %llu done (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:335:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:328:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; sent: %llu done (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:335:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c: In function ‘onscreen_generic’:
/opt/zmap/zmap/src/monitor.c:348:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; send: %llu %sp/s (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:353:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:360:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; send: %llu done (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:365:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c: In function ‘onscreen_generic’:
/opt/zmap/zmap/src/monitor.c:348:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; send: %llu %sp/s (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:353:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c: In function ‘update_status_updates_file’:
/opt/zmap/zmap/src/monitor.c:406:3: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 9 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
   "%s,%u,%u,"
   ^~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:415:3:
   exp->total_sent, exp->send_rate, exp->send_rate_avg,
   ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:408:7: note: format string is defined here
   "%llu,%.0f,%.0f,"
    ~~~^
    %lu
/opt/zmap/zmap/src/monitor.c:406:3: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 15 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
   "%s,%u,%u,"
   ^~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:417:3:
   exp->total_recv, exp->recv_total_rate, exp->recv_total_avg,
   ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:410:7: note: format string is defined here
   "%llu,%.0f,%.0f,"
    ~~~^
    %lu
/opt/zmap/zmap/src/monitor.c:360:29: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 6 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
    "%5s %0.0f%%%s; send: %llu done (%sp/s avg); "
                          ~~~^
                          %lu
/opt/zmap/zmap/src/monitor.c:365:29:
    exp->time_remaining_str, exp->total_sent,
                             ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c: In function ‘update_status_updates_file’:
/opt/zmap/zmap/src/monitor.c:406:3: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 9 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
   "%s,%u,%u,"
   ^~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:415:3:
   exp->total_sent, exp->send_rate, exp->send_rate_avg,
   ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:408:7: note: format string is defined here
   "%llu,%.0f,%.0f,"
    ~~~^
    %lu
/opt/zmap/zmap/src/monitor.c:406:3: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 15 has type ‘uint64_t’ {aka ‘long unsigned int’} [-Wformat=]
   "%s,%u,%u,"
   ^~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:417:3:
   exp->total_recv, exp->recv_total_rate, exp->recv_total_avg,
   ~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/monitor.c:410:7: note: format string is defined here
   "%llu,%.0f,%.0f,"
    ~~~^
    %lu
[ 49%] Building C object src/CMakeFiles/zmap.dir/send.c.o
/opt/zmap/zmap/src/monitor.c: In function ‘monitor_run’:
/opt/zmap/zmap/src/monitor.c:190:56: warning: ‘%s’ directive output may be truncated writing up to 19 bytes into a region of size 18 [-Wformat-truncation=]
   snprintf(exp->time_remaining_str, NUMBER_STR_LEN, " (%s left)",
                                                        ^~
     buf);
     ~~~                                                 
/opt/zmap/zmap/src/monitor.c: In function ‘monitor_run’:
/opt/zmap/zmap/src/monitor.c:190:56: warning: ‘%s’ directive output may be truncated writing up to 19 bytes into a region of size 18 [-Wformat-truncation=]
   snprintf(exp->time_remaining_str, NUMBER_STR_LEN, " (%s left)",
                                                        ^~
     buf);
     ~~~                                                 
In file included from /usr/include/stdio.h:873,
                 from /opt/zmap/zmap/src/monitor.c:16:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: ‘__builtin___snprintf_chk’ output between 9 and 28 bytes into a destination of size 20
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/stdio.h:873,
                 from /opt/zmap/zmap/src/monitor.c:16:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:67:10: note: ‘__builtin___snprintf_chk’ output between 9 and 28 bytes into a destination of size 20
   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        __bos (__s), __fmt, __va_arg_pack ());
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 50%] Building C object src/CMakeFiles/zmap.dir/shard.c.o
[ 51%] Building C object src/CMakeFiles/ztests.dir/recv.c.o
[ 52%] Building C object src/CMakeFiles/ztests.dir/send.c.o
[ 53%] Building C object src/CMakeFiles/zmap.dir/socket.c.o
[ 54%] Building C object src/CMakeFiles/zmap.dir/state.c.o
[ 55%] Building C object src/CMakeFiles/zmap.dir/summary.c.o
/opt/zmap/zmap/src/state.c:45:30: warning: initialization of ‘unsigned int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
       .source_ip_addresses = NULL,
                              ^~~~
/opt/zmap/zmap/src/state.c:45:30: note: (near initialization for ‘zconf.source_ip_addresses[0]’)
/opt/zmap/zmap/src/state.c:13:27: warning: missing braces around initializer [-Wmissing-braces]
 struct state_conf zconf = {.log_level = LOG_INFO,
                           ^
[ 56%] Building C object src/CMakeFiles/ztests.dir/shard.c.o
[ 57%] Building C object src/CMakeFiles/zmap.dir/utility.c.o
[ 58%] Building C object src/CMakeFiles/ztests.dir/state.c.o
[ 58%] Building C object src/CMakeFiles/ztests.dir/socket.c.o
[ 59%] Building C object src/CMakeFiles/ztests.dir/summary.c.o
[ 60%] Building C object src/CMakeFiles/zmap.dir/validate.c.o
/opt/zmap/zmap/src/state.c:45:30: warning: initialization of ‘unsigned int’ from ‘void *’ makes integer from pointer without a cast [-Wint-conversion]
       .source_ip_addresses = NULL,
                              ^~~~
/opt/zmap/zmap/src/state.c:45:30: note: (near initialization for ‘zconf.source_ip_addresses[0]’)
/opt/zmap/zmap/src/state.c:13:27: warning: missing braces around initializer [-Wmissing-braces]
 struct state_conf zconf = {.log_level = LOG_INFO,
                           ^
[ 61%] Building C object src/CMakeFiles/ztests.dir/validate.c.o
[ 62%] Building C object src/CMakeFiles/ztests.dir/ztopt_compat.c.o
[ 63%] Building C object src/CMakeFiles/zmap.dir/zmap.c.o
In file included from /opt/zmap/zmap/src/ztopt_compat.c:17:
/opt/zmap/zmap/src/ztopt.c: In function ‘update_arg’:
/opt/zmap/zmap/src/ztopt.c:345:15: warning: variable ‘val’ set but not used [-Wunused-but-set-variable]
   const char *val = value;
               ^~~
/opt/zmap/zmap/src/ztopt.c:344:9: warning: variable ‘stop_char’ set but not used [-Wunused-but-set-variable]
   char *stop_char = 0;
         ^~~~~~~~~
/opt/zmap/zmap/src/ztopt.c: In function ‘cmdline_parser_internal’:
/opt/zmap/zmap/src/ztopt.c:414:7: warning: variable ‘check_required’ set but not used [-Wunused-but-set-variable]
   int check_required;
       ^~~~~~~~~~~~~~
[ 64%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_icmp_echo.c.o
/opt/zmap/zmap/src/zmap.c: In function ‘main’:
/opt/zmap/zmap/src/zmap.c:464:10: error: ‘struct gengetopt_args_info’ has no member named ‘listen_all_interfaces’
  if (args.listen_all_interfaces) {
          ^
make[2]: *** [src/CMakeFiles/zmap.dir/build.make:283: src/CMakeFiles/zmap.dir/zmap.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:226: src/CMakeFiles/zmap.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 65%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_icmp_echo_time.c.o
[ 66%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_tcp_synscan.c.o
[ 67%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_tcp_synackscan.c.o
[ 69%] Building C object src/CMakeFiles/ztests.dir/probe_modules/packet.c.o
[ 69%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_udp.c.o
[ 70%] Building C object src/CMakeFiles/ztests.dir/probe_modules/probe_modules.c.o
[ 70%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_ntp.c.o
[ 71%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_upnp.c.o
[ 72%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_dns.c.o
In file included from /usr/include/string.h:494,
                 from /opt/zmap/zmap/src/probe_modules/module_ntp.c:13:
In function ‘memcpy’,
    inlined from ‘ntp_init_perthread’ at /opt/zmap/zmap/src/probe_modules/module_ntp.c:211:2:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: warning: ‘__builtin_memcpy’ accessing 90 bytes at offsets 90 and 42 overlaps 42 bytes at offset 90 [-Wrestrict]
   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/zmap/zmap/src/probe_modules/module_ntp.c: In function ‘ntp_make_packet’:
/opt/zmap/zmap/src/probe_modules/module_ntp.c:31:5: warning: function might be candidate for attribute ‘noreturn’ [-Wsuggest-attribute=noreturn]
 int ntp_make_packet(void *buf, ipaddr_n_t src_ip, ipaddr_n_t dst_ip, uint8_t ttl,
     ^~~~~~~~~~~~~~~
[ 73%] Building C object src/CMakeFiles/ztests.dir/probe_modules/module_bacnet.c.o
[ 74%] Building C object src/CMakeFiles/ztests.dir/output_modules/module_csv.c.o
[ 75%] Building C object src/CMakeFiles/ztests.dir/output_modules/module_json.c.o
[ 76%] Building C object src/CMakeFiles/ztests.dir/output_modules/output_modules.c.o
[ 77%] Building C object src/CMakeFiles/ztests.dir/tests/test_harness.c.o
[ 78%] Building C object src/CMakeFiles/ztests.dir/lexer.c.o
[ 79%] Building C object src/CMakeFiles/ztests.dir/parser.c.o
[ 80%] Building C object src/CMakeFiles/ztests.dir/socket-linux.c.o
[ 81%] Building C object src/CMakeFiles/ztests.dir/recv-pcap.c.o
[ 82%] Linking C executable ztests
[ 82%] Built target ztests
make: *** [Makefile:152: all] Error 2
  ```
  
</details>

**Comment by zakird** [comment]
Yeah, I just saw that too. One moment, I think commit got lost.

**Comment by sakul95** [comment]
Now this error occurs:
```
/opt/zmap/zmap/src/zmap.c: In function ‘main’:
/opt/zmap/zmap/src/zmap.c:464:11: error: ‘struct gengetopt_args_info’ has no member named ‘listen_all_interfaces’; did you mean ‘listen_all_interfaces_help’?
  if (args.listen_all_interfaces) {
           ^~~~~~~~~~~~~~~~~~~~~
           listen_all_interfaces_help
make[2]: *** [src/CMakeFiles/zmap.dir/build.make:283: src/CMakeFiles/zmap.dir/zmap.c.o] Error 1
```

**Comment by sakul95** [comment]
Installation had worked, I do get a error from the command itself now when applying the "--listen-all-interfaces" Flag.
```
lukas@machine:/opt/peering-testbed$ sudo zmap -r 100 -M icmp_echoscan -i tap5 --source-ip=184.164.232.1 -I /tmp/hitlisttest_v4 --gateway-mac=c2:39:de:b9:c2:ec --listen-all-interfaces
Dec 18 21:48:45.746 [WARN] zmap: list of IPs is small compared to address space. Performance will suffer, consider using an allowlist instead
Dec 18 21:48:45.746 [INFO] zmap: output module: csv
Dec 18 21:48:45.746 [INFO] csv: no output file selected, will use stdout
Dec 18 21:48:45.763 [INFO] recv: Data link cooked socket
Dec 18 21:48:45.764 [FATAL] recv: couldn't compile filter
```

**Comment by zakird** [comment]
Ah OK, that's unfortunate. It might not be possible to do this looking more at how these generic interfaces are used. I'll read some more, but it doesn't seem optimistic so far.

**Comment by sakul95** [comment]
What a pity, but thanks for your effort! 

If you come up with a solution I will be happy to try it out.

**Comment by zakird** [comment]
I'm going to close this as there doesn't seem to be any good solution to the problem that we can implement.

**Comment by mkoch-tud** [comment]
> Installation had worked, I do get a error from the command itself now when applying the "--listen-all-interfaces" Flag.
> 
> ```
> lukas@machine:/opt/peering-testbed$ sudo zmap -r 100 -M icmp_echoscan -i tap5 --source-ip=184.164.232.1 -I /tmp/hitlisttest_v4 --gateway-mac=c2:39:de:b9:c2:ec --listen-all-interfaces
> Dec 18 21:48:45.746 [WARN] zmap: list of IPs is small compared to address space. Performance will suffer, consider using an allowlist instead
> Dec 18 21:48:45.746 [INFO] zmap: output module: csv
> Dec 18 21:48:45.746 [INFO] csv: no output file selected, will use stdout
> Dec 18 21:48:45.763 [INFO] recv: Data link cooked socket
> Dec 18 21:48:45.764 [FATAL] recv: couldn't compile filter
> ```

I just ran into the same issue -- i looked at the content of the bpf_filter which is:
`not ether src [HW-mac] (icmp6 && (ip6[40] == 129 || ip6[40] == 3 || ip6[40] == 1 || ip6[40] == 2 || ip6[40] == 4) && ip6 dst host [IP]`
(I'm using the IPv6 version of zmap, therefore icmp6 and ip6)
When I remove `not ether src [HW-mac]` it works perfectly fine.
My intuition says that it makes sense not filtering for HW Mac when listening on all interfaces. However, I might misunderstand how the bpf filter works.
The fix is to only add this part of the bpf filter when not listening on all interfaces in `recv-pcap.c` -- simply add `&& !zconf.listen_all_interfaces` for two `if` conditions:
```
if (!zconf.send_ip_pkts && !zconf.listen_all_interfaces) {
                snprintf(bpftmp, sizeof(bpftmp) - 1,
                         "not ether src %02x:%02x:%02x:%02x:%02x:%02x",
                         zconf.hw_mac[0], zconf.hw_mac[1], zconf.hw_mac[2],
                         zconf.hw_mac[3], zconf.hw_mac[4], zconf.hw_mac[5]);
                assert(strlen(zconf.probe_module->pcap_filter) + 10 <
                       (BPFLEN - strlen(bpftmp)));
        } else {
                bpftmp[0] = 0;
        }
        if (zconf.probe_module->pcap_filter) {
                if (!zconf.send_ip_pkts && !zconf.listen_all_interfaces) {
                        strcat(bpftmp, " and (");
                } else {
                        strcat(bpftmp, "(");
                }
                strcat(bpftmp, zconf.probe_module->pcap_filter);
                strcat(bpftmp, ")");
                log_info("bpf filter",bpftmp);
        }

```

**Comment by mkoch-tud** [comment]
@zakird @dadrian if you reopen this issue i can assign a PR

**Comment by zakird** [comment]
@mkoch-tud, If there's a solution that works, I'm more than happy to accept a PR. I think the original MAC-based filter was to exclude the packets that ZMap was sending out from being processed by the receive code. Given the scale of probes we send out, I suspect that's a pretty critical exclusion.

**Comment by mkoch-tud** [comment]
If i understood it correctly the BPF filter fails to compile because in 'cooked mode' (capturing on ANY interface) is done by reading from a PF_PACKET/SOCK_DGRAM socket rather than the PF_PACKET/SOCK_RAW socket normally used for capturing. That means no link-layer header is available. (see https://wiki.wireshark.org/SLL )
In that case using `sll.src.eth` instead of `ether src` might be an option (https://www.wireshark.org/docs/dfref/s/sll.html).
I'll try it out asap

**Comment by mkoch-tud** [comment]
Using `sll` was nonsense.
Would it be an option to use an upper layer protocol for filtering, i.e. filter out packets with source IP address we use as source address for scanning?

**Comment by zakird** [comment]
I suspect yes, but we need to support the fact that we send from multiple IP addresses.

**Comment by mkoch-tud** [comment]
To come back on this topic -- I decided to go for a different approach as listening on ANY turns out to be absolutely painful to maintain. 
I added another parameter named `--recviface` as a second interface and changed the `recv-pcap.c` to use this iface for receiving packets.
Then, I setup a dummy interface (`ip link add dummy01 type dummy && ip link set dummy01 up`) on my machine and configured a static route using `ip route add [ip range] dev dummy01` that routes all traffic destined for my source ip range to that interface. 
This works perfectly but it's not a nice fix for a PR as you still need to play around with your routing table manually.
However, it might help others who stumbled across the same issue.
